pkg/patch2/patch_test.go

```diff
@@ -0,0 +1,738 @@
+				key := client.ObjectKey{Name: obj.Name, Namespace: obj.Namespace}
+				defer func() {
+					Expect(ctx.Client.Delete(ctx, obj)).To(Succeed())
+				}()
+
+				By("Creating a new patch helper")
+				patcher, err := NewHelper(obj, ctx.Client)
+				Expect(err).NotTo(HaveOccurred())
+
+				By("Updating the object spec")
+				obj.Spec.ImageName = "image-name-42"
+
+				By("Patching the object")
+				Expect(patcher.Patch(ctx, obj)).To(Succeed())
+
+				By("Validating the object has been updated")
+				Eventually(func() bool {
+					objAfter := obj.DeepCopy()
+					if err := ctx.Client.Get(ctx, key, objAfter); err != nil {
+						return false
+					}
+
+					return obj.Spec.ImageName == objAfter.Spec.ImageName
+				}, timeout).Should(BeTrue())
+			})
+
+			Specify("updating status", func() {
+				obj := obj.DeepCopy()
+
+				By("Creating the object")
+				Expect(ctx.Client.Create(ctx, obj)).ToNot(HaveOccurred())
+				key := client.ObjectKey{Name: obj.Name, Namespace: obj.Namespace}
+				defer func() {
+					Expect(ctx.Client.Delete(ctx, obj)).To(Succeed())
+				}()
+
+				By("Creating a new patch helper")
+				patcher, err := NewHelper(obj, ctx.Client)
+				Expect(err).NotTo(HaveOccurred())
+
+				By("Updating the object status")
+				obj.Status.Host = "vm-host"
+
+				By("Patching the object")
+				Expect(patcher.Patch(ctx, obj)).To(Succeed())
+
+				By("Validating the object has been updated")
+				Eventually(func() bool {
+					objAfter := obj.DeepCopy()
+					if err := ctx.Client.Get(ctx, key, objAfter); err != nil {
+						return false
+					}
+					return reflect.DeepEqual(objAfter.Status, obj.Status)
+				}, timeout).Should(BeTrue())
+			})
+
+			Specify("updating both spec, status, and adding a condition", func() {
+				obj := obj.DeepCopy()
+				obj.ObjectMeta.Namespace = ctx.Namespace
+
+				By("Creating the object")
+				Expect(ctx.Client.Create(ctx, obj)).ToNot(HaveOccurred())
+				key := client.ObjectKey{Name: obj.Name, Namespace: obj.Namespace}
+				defer func() {
+					Expect(ctx.Client.Delete(ctx, obj)).To(Succeed())
+				}()
+
+				By("Creating a new patch helper")
+				patcher, err := NewHelper(obj, ctx.Client)
+				Expect(err).NotTo(HaveOccurred())
+
+				By("Updating the object spec")
+				obj.Spec.ImageName = "image-name"
+
+				By("Updating the object status")
+				obj.Status.Host = "vm-host"
+
+				By("Setting Ready condition")
+				conditions.MarkTrue(obj, readyConditionType)
+
+				By("Patching the object")
+				Expect(patcher.Patch(ctx, obj)).To(Succeed())
+
+				By("Validating the object has been updated")
+				Eventually(func() bool {
+					objAfter := obj.DeepCopy()
+					if err := ctx.Client.Get(ctx, key, objAfter); err != nil {
+						return false
+					}
+
+					return obj.Status.Host == objAfter.Status.Host &&
+						conditions.IsTrue(objAfter, readyConditionType) &&
+						reflect.DeepEqual(obj.Spec, objAfter.Spec)
+				}, timeout).Should(BeTrue())
+			})
+		})
+
+		/*
```

Comment:
remove the commented out tests? . what plans for this?

Reply:
These are commented out in existing `pkg/patch` tests from CAPI, likely because our types didn't have an observed generation field. I'm trying to keep these diffs small at the moment to just get us close to v1a1 feature parity w/ v1a2. metav1.Conditions do have a generation field so these tests are something that we can work on reenabling. I've added this work to my internal P2 tracking stories.